### ISSUE TITLE
Error while running appcenter_upload . bad URI(is not URI?)

### ISSUE THREAD

**Issue by rkhandelwal17** [microsoft/fastlane-plugin-appcenter#152]
App Center has the URL formatted like this:
https://appcenter.ms/users/myname/apps/My-App-Name
But the value generated in fastlane is:

/v0.1/apps/mynamehere/My App Name

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.14.5                                     |
| Ruby                        | 2.3.8                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.17.2 (Apple Git-113)          |
| Installation Source         | /usr/local/lib/ruby/gems/2.3.0/bin/fastlane |
| Host                        | Mac OS X 10.14.5 (18F132)                   |
| Ruby Lib Dir                | /usr/local/Cellar/ruby@2.3/2.3.8_1/lib      |
| OpenSSL Version             | OpenSSL 1.0.2q  20 Nov 2018                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 10.1                                        |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_AU.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
opt_out_usage

fastlane_version "2.109.1"
xcversion(version: "10.1")

default_platform :ios

# Environment
artifactory_endpoint = "https://artifactory.srv.westpac.com.au/artifactory"
carthage_repo = "A00931_Carthag"
#ipa_path = "Test/Snapshot/au/com/westpac/native/Banking.Universal.iOS/#{last_git_commit[:abbreviated_commit_hash]}/"
#new_framework_path = "#{prebuild_options[:package_path]}/#{new_ver}/
#ipa_path = "#{artifactory_endpoint}/#{carthage_repo}/au/com/westpac/native/Banking.Universal.ios"

type_mapping = {
  "develop" => { folder: "Develop", pre_type: "alpha" },
  "snapshot" => { folder: "Snapshot", pre_type: "alpha" },
  "beta" => { folder: "Beta", pre_type: "beta" },
  "release" => { folder: "Release", pre_type: nil },
  "unstable" => { folder: "Unstable", pre_type: "alpha" }
}

platform :ios do

  def find_keychain_password( key: nil )
    if !key.nil?
      item = Security::InternetPassword.find(account: key)
      if !item.nil?
        item.password
      end
    end
  end

  before_all do |lane|
    if lane == :alpha or lane == :beta_debug or lane == :beta_uat or lane == :release
      $artifactory_api_key = find_keychain_password( key: "artifactory_api_key" )
      if $artifactory_api_key.nil?
        UI.abort_with_message!("
          \r***** Missing \"artifactory_api_key\" credential
          \rSave this credential in the keychain using fastlane fastlane-credentials add
          ")
      end
    end
  
    if lane == :push_to_hockey or lane == :push_to_hockey_beta or lane == :push_to_hockey_uat or lane == :push_to_hockey_prod
      $hockey_api_token = find_keychain_password( key: "hockey_api_token" )
      if $hockey_api_token.nil?
        UI.abort_with_message!("
          \r***** Missing \"hockey_api_token\" credential
          \rSave this credential in the keychain using fastlane fastlane-credentials add
          ")
      end
    end

    $is_bamboo = false
    $is_merge_gate = false
    $working_branch = nil
    $merge_branch = nil
    
    bamboo_branch = ENV["bamboo_planRepository_1_branch"]
    bamboo_integration_branch = ENV["bamboo_planRepository_integrationBranch_branch"]
    git_branch_result = git_branch

    if !bamboo_branch.nil? and !bamboo_branch.empty?
      $working_branch = bamboo_branch
      $is_bamboo = true
      if !bamboo_integration_branch.nil?
        $is_merge_gate = true
        $merge_branch = bamboo_integration_branch
      end
    elsif !git_branch_result.nil? and !git_branch_result.empty?
      $working_branch = git_branch_result
    else
      UI.abort_with_message!("
        \r***** Could not identify the current branch and will be unable to direct script execution.
        \rYou are probably running this locally and have entered a \"detached HEAD\" state.")
    end
    UI.message("Working branch identified as: #{$working_branch}")

    if $is_merge_gate
      UI.message("Merge branch identified as: #{$merge_branch}")
    end
  end

  lane :cartwright_dev do
    path = File.expand_path("~/Library/Caches/carthage")
    if File.directory?(path)
      UI.message("Removing Carthage HTTP cache from: #{path}")
      FileUtils.remove_dir(path)
    end

    cartwright(
      remote: "https://artifactory.srv.westpac.com.au/artifactory/A00931_Carthag/Snapshot/cartwright/1.0.0/Cartwright"
    )
    carthage(
      command: "update"
    )
  end

  lane :cartwright_beta do
    path = File.expand_path("~/Library/Caches/carthage")
    if File.directory?(path)
      UI.message("Removing Carthage HTTP cache from: #{path}")
      FileUtils.remove_dir(path)
    end

    cartwright(
      remote: "https://artifactory.srv.westpac.com.au/artifactory/A00931_Carthag/Beta/cartwright/1.0.0/Cartwright"
    )
    carthage(
      command: "update"
    )
  end

  lane :cartwright_unstable do
    cartwright(
      remote: "https://artifactory.srv.westpac.com.au/artifactory/A00931_Carthag/Unstable/cartwright/0.1/Cartwright"
    )
    carthage(
      command: "update"
    )
  end
  
  # options expects :build_type
  private_lane :prebuild do |options|
  
    # get details of repo
    project_version = get_version_number(xcodeproj: "Banking.xcodeproj")
    bundle_id = "au.com.westpac.native.BankingEnterpriseDebugDistribution"
    bundle_components = bundle_id.split('.') # e.g. ["au", "com", "westpac", "native", "BankingEnterpriseDebugDistribution"]
    package_name = bundle_components[-1] # e.g. "BankingEnterpriseDebugDistribution"

    values = {}
    values[:project_version] = project_version
    values[:bundle_id] = bundle_id
    values[:bundle_components] = bundle_components
    values[:package_name] = package_name

    values[:commit_hash] = last_git_commit[:abbreviated_commit_hash]

    UI.message("Project version read as: #{project_version}")
    UI.message("Bundle ID read as: #{bundle_id}")

    begin
      Semantic::Version.new(project_version)
    rescue ArgumentError => e
      UI.user_error!("Could not parse version from project: #{e.message}")
    end
    
    # package folder path in artifactory
    manifest_components = []
    if $FASTLANE_TEST_MODE
      manifest_components.push("Test")
    end
    type_folder = type_mapping[options[:build_type]][:folder] # e.g. Snapshot
    manifest_components.push(type_folder)
    manifest_components.concat(bundle_components) # e.g. ["Test", "Snapshot" ,"au", "com", "westpac", "native", "BankingEnterpriseDebugDistribution"]
    values[:package_path] = manifest_components.join("/") # e.g. "Test/Snapshot/au/com/westpac/native/BankingEnterpriseDebugDistribution"
    
    # JSON manifest path in artifactory
    manifest_name = "#{package_name}.json"
    values[:manifest_path] = "#{values[:package_path]}/#{manifest_name}" # e.g. "Test/Snapshot/au/com/westpac/native/WBGFoundation/WBGFoundation.json"
    values[:manifest_name] = manifest_name

    values # return object

  end
  
  # options expects :prebuild_options and :build_type
  private_lane :postbuild do |options|
    prebuild_options = options[:prebuild_options]

    UI.message("Attempting to download carthage manifest")
    manifest_download = artifactory_download_anon(
      endpoint: artifactory_endpoint,
      repo: carthage_repo,
      repo_path: prebuild_options[:manifest_path],
      file: prebuild_options[:manifest_name]
    )

    if manifest_download.nil?
      # 404 error and file won't exist
      manifest = {}
      UI.important("Could not find an existing manifest for this package! Creating a new one instead.")
    else
      manifest = JSON.parse(File.read(manifest_download))
      UI.message("Manifest downloaded and parsed successfully")
    end

    # look up plist version against keys in the manifest hash.
    # If found parse URL for existing version and update, otherwise create new.
    project_version = prebuild_options[:project_version]
    current_url = manifest[project_version]

    pre_type = type_mapping[options[:build_type]][:pre_type]
    if pre_type.nil?
      if current_url.nil?
        # good outcome - nothing exists for this version.
        new_ver = Semantic::Version.new(project_version)
      else
        UI.user_error!("#{type_mapping[options[:build_type]][:folder]} build for version #{project_version} already existis in Carthage JSON! Will not overwright version.")
      end
    else
      if !current_url.nil?
        # grab last version with the same major/minor/patch version
        last_ver = Semantic::Version.new(current_url.split("/")[-2])

        # increment pre-release version number on the last element
        new_pre_arr = last_ver.pre.split(".")
        new_pre_arr[-1] = new_pre_arr[-1].to_i.next
        new_pre = new_pre_arr.join(".")

        # clone the existing version and set pre and hash
        new_ver = last_ver.dup
        new_ver.pre = new_pre
      else
        new_ver = Semantic::Version.new(prebuild_options[:project_version])
        new_ver.pre = "#{pre_type}.1"
      end
    end

    new_ver.build = prebuild_options[:commit_hash]

    UI.message("SemVer string for this snapshot is #{new_ver}")

    package_name = prebuild_options[:package_name]
    package_with_version = "#{package_name}_#{new_ver}" # e.g. "BankingEnterpriseDebugDistribution_0.1.0-alpha.1+abc567"

    # rename ipa to include the version
    original_ipa = "#{package_name}.ipa" # e.g. new_app_path"BankingEnterpriseDebugDistribution.ipa.zip"
    new_ipa = "#{package_with_version}.ipa" # e.g. "BankingEnterpriseDebugDistribution_0.1.0-alpha.1+abc567.ipa.zip"
    FileUtils.cp("../#{original_ipa}", "../#{new_ipa}")

    # rename dsym to include the version
    original_dsym = "#{package_name}.app.dSYM.zip" # e.g. "BankingEnterpriseDebugDistribution.app.dSYM.zip"
    new_dsym = "#{package_with_version}.app.dSYM.zip" # e.g. "BankingEnterpriseDebugDistribution_0.1.0-alpha.1+abc567.app.dSYM.zip"
    FileUtils.cp("../#{original_dsym}", "../#{new_dsym}")
    
    # modify and save the manifest with the new snapshot
    new_ipa_path = "#{prebuild_options[:package_path]}/#{new_ver}/#{new_ipa}"
    manifest[project_version] = "#{artifactory_endpoint}/#{carthage_repo}/#{new_ipa_path}"
    
    new_dsym_path = "#{prebuild_options[:package_path]}/#{new_ver}/#{new_dsym}"
    manifest[project_version] = "#{artifactory_endpoint}/#{carthage_repo}/#{new_dsym_path}"

    new_carthage_manifest = File.open("../#{package_name}.json", "w+")
    new_carthage_manifest.puts(JSON.generate(manifest))
    new_carthage_manifest.close

    # publish snapshot to the artifactory
    artifactory_with_key(
      file: new_ipa,
      repo: carthage_repo,
      repo_path: new_ipa_path,
      endpoint: artifactory_endpoint,
      api_key: read_keychain(key: "artifactory_api_key")
    )

    # publish dsym to the artifactory
    artifactory_with_key(
      file: new_dsym,
      repo: carthage_repo,
      repo_path: new_dsym_path,
      endpoint: artifactory_endpoint,
      api_key: read_keychain(key: "artifactory_api_key")
    )

  end

  # --------------------------------------------------------------------------------
  # generate hockey release notes based on comments from the last merge/commit
  # include commit hash and recent commit messages

  lane :hockey_release_notes do |options|
    # commit hash
    short_hash = last_git_commit[:abbreviated_commit_hash] # short sha of commit

    # commit changes
    commit_change = changelog_from_git_commits(
      between: ["HEAD~1", "HEAD"],
      pretty: "- %s"
      )
    
    # SCP changelogs for the Framework/Features
       scp(
       host: "10.27.30.112",
       username: "mobadmin",
       download: {
           src: "Changelogs.txt",
           dst: "."
         }
      )
    
    scp_changelogs = Helper.backticks("cat ../Changelogs.txt")
    UI.message("***** scp_changelogs=#{scp_changelogs}")
 
    # release notes: in ruby, the last assignment value is returned to the caller
    release_notes = "Commit: #{short_hash} #{commit_change}\n\n #{scp_changelogs}"
  end

  lane :alpha do
    fastlane_require 'artifactory'
    fastlane_require 'semantic'
    require 'json'
    require 'fileutils'

    prebuild_options = prebuild(
      build_type: "snapshot"
    )

    cartwright_dev

    gym(
      scheme: "BankingEnterpriseDebugDistribution",
      export_method: "enterprise",
      codesigning_identity: "iPhone Distribution: WESTPAC BANKING CORPORATION",
      derived_data_path: "./derived_data/",
      build_path: "./derived_data/",
      export_options: { compileBitcode: false,stripSwiftSymbols: false },
      buildlog_path: ".",
      clean: true,
      configuration: "DebugDistribution",
      output_name: "BankingEnterpriseDebugDistribution"
    )
    clear_derived_data(derived_data_path: "./derived_data/")

    postbuild(
      prebuild_options: prebuild_options,
      build_type: "snapshot"
    )

  end  

  lane :push_to_hockey do
    appcenter_upload(
      #app_name: "WBC-Beautiful-App-Phone-Test",
      ipa: "./BankingEnterpriseDebugDistribution.ipa",
      dsym: "./BankingEnterpriseDebugDistribution.app.dSYM.zip",
      api_token: "..token..",
      #api_token: $hockey_api_token,
      owner_name: "Westpac Mobile App Development",
      #public_identifier: "8073b81869304bcd856c6198d8089797",
      #release_type: "2",
      #timeout: 120,
      release_notes: hockey_release_notes
      #bypass_cdn: true
    )
  end

  lane :beta_debug do
    fastlane_require 'artifactory'
    fastlane_require 'semantic'
    require 'json'
    require 'fileutils'

    prebuild_options = prebuild(
      build_type: "beta"
    )

    cartwright_beta

    gym(
      scheme: "BankingEnterpriseDebugDistribution",
      export_method: "enterprise",
      codesigning_identity: "iPhone Distribution: WESTPAC BANKING CORPORATION",
      derived_data_path: "./derived_data/",
      build_path: "./derived_data/",
      export_options: { compileBitcode: false,stripSwiftSymbols: false },
      buildlog_path: ".",
      clean: true,
      configuration: "DebugDistribution",
      output_name: "BankingEnterpriseDebugDistribution"
    )
    clear_derived_data(derived_data_path: "./derived_data/")

    postbuild(
      prebuild_options: prebuild_options,
      build_type: "beta"
    ) 
  end
  
  lane :push_to_hockey_beta do  
    hockey(
      ipa: "BankingEnterpriseDebugDistribution.ipa",
      dsym: "BankingEnterpriseDebugDistribution.app.dSYM.zip",
      api_token: $hockey_api_token,
      public_identifier: "5f0ee20d345143ef8d3599507a997ced",
      release_type: "2",
      timeout: 120,
      notes: hockey_release_notes,
      bypass_cdn: true
    )
  end

  lane :beta_uat do
    fastlane_require 'artifactory'
    fastlane_require 'semantic'
    require 'json'
    require 'fileutils'

    prebuild_options = prebuild(
      build_type: "beta"
    )
  
    cartwright_beta

    gym(
      scheme: "BankingEnterpriseUAT",
      export_method: "enterprise",
      codesigning_identity: "iPhone Distribution: WESTPAC BANKING CORPORATION",
      derived_data_path: "./derived_data/",
      build_path: "./derived_data/",
      export_options: { compileBitcode: false,stripSwiftSymbols: false },
      buildlog_path: ".",
      clean: true,
      configuration: "UAT",
      output_name: "BankingEnterpriseUAT"
    )
    clear_derived_data(derived_data_path: "./derived_data/")
    
    postbuild(
      prebuild_options: prebuild_options,
      build_type: "beta"
    ) 
  end 
    
  lane :push_to_hockey_uat do
    hockey(
      ipa: "BankingEnterpriseUAT.ipa",
      dsym: "BankingEnterpriseUAT.app.dSYM.zip",
      api_token: $hockey_api_token,
      public_identifier: "8073b81869304bcd856c6198d8089797",
      release_type: "2",
      timeout: 120,
      notes: hockey_release_notes,
      bypass_cdn: true
    )
  end

  lane :release do
    fastlane_require 'artifactory'
    fastlane_require 'semantic'
    require 'json'
    require 'fileutils'

    prebuild_options = prebuild(
      build_type: "release"
    )

    cartwright_beta

    gym(
      scheme: "Banking",
      export_method: "enterprise",
      codesigning_identity: "iPhone Distribution: WESTPAC BANKING CORPORATION",
      derived_data_path: "./derived_data/",
      build_path: "./derived_data/",
      export_options: { compileBitcode: false,stripSwiftSymbols: false },
      buildlog_path: ".",
      clean: true,
      configuration: "Release",
      output_name: "BankingEnterprise"
    )
    clear_derived_data(derived_data_path: "./derived_data/")

    postbuild(
      prebuild_options: prebuild_options,
      build_type: "release"
    ) 
  end 

  lane :push_to_hockey_prod do
    hockey(
      ipa: "BankingEnterprise.ipa",
      dsym: "BankingEnterprise.app.dSYM.zip",
      api_token: $hockey_api_token,
      public_identifier: "5ab95e8c2fa94f4c98b8a500ad1f6d04",
      release_type: "2",
      timeout: 120,
      bypass_cdn: true
    )
  end

end

```
</details>

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.134.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                    | Version | Update-Status |
| ------------------------- | ------- | ------------- |
| fastlane-plugin-appcenter | 1.6.0   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.0.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| CFPropertyList            | 3.0.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.6.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-spinner               | 0.9.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20180417 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday                   | 0.17.0       |
| faraday_middleware        | 0.13.1       |
| fastimage                 | 2.1.5        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.9.5        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| bundler                   | 1.17.3       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| jwt                       | 2.1.0        |
| uber                      | 0.1.0        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mime-types                | 3.2.2        |
| multi_json                | 1.13.1       |
| signet                    | 0.11.0       |
| memoist                   | 0.16.0       |
| googleauth                | 0.6.7        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.23.9       |
| google-cloud-env          | 1.0.5        |
| google-cloud-core         | 1.3.0        |
| digest-crc                | 0.4.1        |
| google-cloud-storage      | 1.16.0       |
| tty-cursor                | 0.6.1        |
| tty-screen                | 0.7.0        |
| json                      | 2.2.0        |
| io-console                | 0.4.5        |
| excon                     | 0.64.0       |
| mini_portile2             | 2.4.0        |
| nokogiri                  | 1.10.1       |
| atomos                    | 0.1.3        |
| claide                    | 1.0.2        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.9.0        |
| unicode-display_width     | 1.6.0        |
| os                        | 1.0.1        |
| psych                     | 2.1.0.1      |
| mime-types-data           | 3.2019.0331  |
| fastlane-plugin-appcenter | 1.6.0        |
</details>

*generated on:* **2019-11-11**
</details>

**Comment by Raes** [comment]
I got the same message when I did not configure my properties correctly. Would suggest following the README and adding variables to your ENV file. In App Center click on the app you are looking to upload to and check out the URL to get there. In my case "owner_name" was not what I expected, neither was "app_name"

**Comment by Ajaykn21** [comment]
Hi @rkhandelwal17 
Yeah we have the AppCenter URL format in the portal is different from the URL we use for Api's (also fastlane).
Portal: https://appcenter.ms/users/<owner_name>/apps/<app_name>
API/fastlane: /v0.1/apps/<owner_name>/<app_name>

@Raes : Note that the owner_name is the AppCenter generated owner name and same goes for app name as well. 

These shouldn't be an issue, as long as you provide correct owner_name and app_name in the properties.

**Comment by Ajaykn21** [comment]
@rkhandelwal17 , the issue has been identified for the error 
` appcenter_upload . bad URI(is not URI?)`
This is because you might be giving the appName with spaces in it and so its failing.
We will work on this.

To unblock yourself, please give the app_name without spaces.
And if you are looking for creating the app using the plugin, you can set the app_display_name to whatever the friendly name you want.

**Comment by Ajaykn21** [comment]
`appcenter_upload . bad URI(is not URI?)` is fixed and released in v1.7.0
